### PR TITLE
Refactor rocker dependencies into java Feature API (partial)

### LINKED ISSUE #1681: Generated projects with `google-cloud-function-http` feature include no-longer-needed dependencies
The project template generated when a project includes the `google-cloud-function-http` feature includes the following comment:

https://github.com/micronaut-projects/micronaut-starter/blob/5a81642955e30defcb5f7c602f3b93e4fd1e7a1b/starter-core/src/main/java/io/micronaut/starter/feature/build/maven/templates/pom.rocker.raw#L104

https://github.com/GoogleCloudPlatform/functions-framework-java/pull/32 has long since been merged and I think the additional dependencies can be removed. I'd be happy to open a pull request, but it's not entirely clear to me which of the dependencies (just the dependency following the comment? all of the dependencies following the comment?) are affected.

### PR DESCRIPTION
**Pull request by wetted**
On Issue #1698 @sdelamo says, "we should move the logic of adding the dependencies outside of rocker into Java classes."

This implements a large portion of that. What's still left is somewhat more complicated – with conditional logic between features, inconsistencies between gradle and maven, and for other reasons – but there's no reason that should hold this part up.

The dependency logic for the azure features is broken for gradle. That's a bug so I will fix that separately.

### CODE REVIEW

**Comment by sdelamo** [comment]
@wetted is this ready for review? it is still in draft

**Comment by wetted** [comment]
> @wetted is this ready for review? it is still in draft

@sdelamo no, the unit tests pass, but something is broken in the functional/integration tests I need to address.

**Comment by wetted** [comment]
I am working the Azure, GCP, and OracleCloud Rocker template refactors in separate PRs. They are somewhat more complicated, and have bugs for Gradle projects.

**Comment by wetted** [comment]
> @wetted is this ready for review? it is still in draft

@sdelamo this is good to go now.

File: starter-core/src/main/java/io/micronaut/starter/feature/function/awslambda/AwsLambda.java [hunk 1]
```diff
@@ -246,6 +251,12 @@ private void addDependencies(@NonNull GeneratorContext generatorContext) {
         if (generatorContext.hasFeature(AwsLambdaSnapstart.class)) {
             generatorContext.addDependency(DEPENDENCY_MICRONAUT_CRAC);
         }
+
+        if (generatorContext.getFeatures().testFramework().isSpock() &&
+                generatorContext.getBuildTool().isGradle()) {
+            // maven has this in parent pom
+            generatorContext.addDependency(DEPENDENCY_MICRONAUT_FUNCTION_TEST);
+        }
```
Comment (alvarosanchez) [inline starter-core/src/main/java/io/micronaut/starter/feature/function/awslambda/AwsLambda.java:259]: Would it make sense to move this to the parent `AwsCloudFeature`?

File: starter-core/src/main/java/io/micronaut/starter/feature/function/gcp/GoogleCloudFunction.java [hunk 2]
```diff
@@ -63,6 +64,11 @@ public boolean isVisible() {
         return false;
     }
 
+    @Override
+    public void apply(GeneratorContext generatorContext) {
+        super.apply(generatorContext);
+    }
```
Comment (alvarosanchez) [inline starter-core/src/main/java/io/micronaut/starter/feature/function/gcp/GoogleCloudFunction.java]: Is this needed?

File: starter-core/src/main/java/io/micronaut/starter/feature/other/OpenApi.java [hunk 3]
```diff
@@ -56,6 +58,14 @@ public void apply(GeneratorContext generatorContext) {
                 .groupId("io.swagger.core.v3")
                 .artifactId("swagger-annotations")
                 .compile());
+        if (generatorContext.getBuildTool() == BuildTool.MAVEN &&
+                generatorContext.getLanguage() == Language.GROOVY) {
+            generatorContext.addDependency(Dependency.builder()
+                    .groupId("io.micronaut.openapi")
+                    .artifactId("micronaut-openapi")
+                    .versionProperty("micronaut.openapi.version")
```
Comment (alvarosanchez) [inline starter-core/src/main/java/io/micronaut/starter/feature/other/OpenApi.java]: This is probably in the BOM, so the version should be needed

**Review by alvarosanchez — CHANGES_REQUESTED** [review]

File: starter-core/src/main/java/io/micronaut/starter/feature/function/awslambda/AwsLambda.java (hunk 1 above)
Comment (wetted) [reply starter-core/src/main/java/io/micronaut/starter/feature/function/awslambda/AwsLambda.java:259]: @alvarosanchez I don't think we want to put it there. `AwsCloudFeature` is an interface that extends `CloudFeature` and not `Feature`. What I have now still matches the logic that was in the Rocker template.

**Comment by wetted** [comment]
@alvarosanchez @sdelamo 

I added another change set that refactors the GCP dependencies from rocker files into the Feature classes. Please review that as well.

**Review by alvarosanchez — APPROVED** [review]

**Review by alvarosanchez — APPROVED** [review]

**Comment by wetted** [comment]
I am now able to duplicate the failing test locally. I am working on fixing it.

**Pull request merged**